File: components/file/local/file_test.go
/*
 * Copyright 2021 Layotto Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package local

import (
	"context"
	"io"
	"os"
	"strconv"
	"testing"

	"github.com/stretchr/testify/assert"

	"mosn.io/layotto/components/file"
)

const (
	FileName = "test.txt"
)

func WriteFile(writer *io.PipeWriter) {
	writer.Write([]byte("hello"))
	writer.Close()
}

func CheckFileExist(name string) (bool, error) {
	_, err := os.Stat(name)
	if err == nil {
		return true, nil
	}
	if os.IsNotExist(err) {
		return false, nil
	}
	return false, nil
}

func TestFile(t *testing.T) {
	ls := &LocalStore{}
	f := &file.PutFileStu{}
	f.FileName = FileName
	f.Metadata = make(map[string]string)
	reader, writer := io.Pipe()
	err := ls.Put(context.TODO(), f)
	assert.Equal(t, err.Error(), "fileMode is required for put file")
	mode := 0777
	f.Metadata[FileMode] = strconv.Itoa(mode)
	err = ls.Put(context.TODO(), f)
	assert.Equal(t, err.Error(), "fileFlag is required for put file")
	f.Metadata[FileFlag] = strconv.Itoa(os.O_RDWR | os.O_CREATE)
	f.DataStream = reader
	go WriteFile(writer)
	err = ls.Put(context.TODO(), f)
	assert.Nil(t, err)
	exist, err := CheckFileExist(f.FileName)
	assert.Nil(t, err)
	assert.Equal(t, true, exist)

	data := make([]byte, 10)
	fs := &file.GetFileStu{}
	fs.FileName = FileName
	stream, err := ls.Get(context.TODO(), fs)
	assert.Nil(t, err)
	n, err := stream.Read(data)
	stream.Close()
	assert.Nil(t, err)
	assert.Equal(t, string(data[:n]), "hello")

	fr := &file.ListRequest{}
	fr.DirectoryName = "./"
	res, err := ls.List(context.TODO(), fr)
	assert.Nil(t, err)
	assert.Equal(t, len(res.Files), 3)
	for _, v := range res.Files {
		t.Log(v)
	}

	fr.Marker = "file.go"
	fr.PageSize = 1
	res, err = ls.List(context.TODO(), fr)
	assert.Nil(t, err)
	assert.Equal(t, res.IsTruncated, true)
	assert.Equal(t, res.Marker, "file_test.go")
	assert.Equal(t, len(res.Files), 1)
	for _, v := range res.Files {
		t.Log(v)
	}

	st := &file.FileMetaRequest{FileName: "hello.txt"}
	r, err := ls.Stat(context.TODO(), st)
	assert.Equal(t, file.ErrNotExist, err)
	assert.Nil(t, r)

	st.FileName = "file.go"
	r, err = ls.Stat(context.TODO(), st)
	assert.Nil(t, err)
	assert.Equal(t, r.Metadata[FileIsDir][0], "false")

	fd := &file.DelRequest{}
	fd.FileName = FileName
	err = ls.Del(context.TODO(), fd)
	assert.Nil(t, err)
	exist, err = CheckFileExist(FileName)
	assert.Nil(t, err)
	assert.Equal(t, exist, false)
}

File: sdk/js-sdk/test/unit/client/File.test.ts
/*
 * Copyright 2021 Layotto Authors
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { strict as assert } from 'assert';
import { tmpdir } from 'os';
import { existsSync, createWriteStream, createReadStream } from 'fs';
import { mkdtemp, rm } from 'fs/promises';
import { pipeline } from 'stream/promises';
import { join } from 'path';
import { Client, utils } from '../../../src';

describe.skip('client/File.test.ts', () => {
  let client: Client;
  let tmpfileDir: string;
  const storeName = 'aliOSS';
  const bucket = 'layotto-js-sdk-local-test';

  beforeAll(async () => {
    client = new Client();
    tmpfileDir = await mkdtemp(join(tmpdir(), 'layotto-js-sdk-file-unittest-tmp-'));
  });

  afterAll(async () => {
    // rm -rf
    await rm(tmpfileDir, { recursive: true, force: true });
  });

  describe('get()', () => {
    it('should get a file stream success', async () => {
      const filepath = join(tmpfileDir, 'getFileStream.jpg');
      console.log(filepath);
      assert(!existsSync(filepath));

      const stream = await client.file.get({
        storeName: 'aliOSS',
        name: 'foo/tab3.jpg',
      });
      assert(stream);
      await pipeline(
        stream,
        createWriteStream(filepath),
      );
      assert(existsSync(filepath));
    }, 30000);

    it('should get a file stream from minio', async () => {
      const filepath = join(tmpfileDir, 'minio.jpg');
      console.log(filepath);
      assert(!existsSync(filepath));

      const stream = await client.file.get({
        storeName: 'minioOSS',
        name: '1.jpg',
        metadata: {
          bucket,
          endpoint: 'http://127.0.0.1:9000',
        },
      });
      assert(stream);
      await pipeline(
        stream,
        createWriteStream(filepath),
      );
      assert(existsSync(filepath));
    });

    it('should throw error when storeName not exists', async () => {
      await assert.rejects(
        async () => {
          const filepath = join(tmpfileDir, 'storeName-not-exists.jpg');
          const stream = await client.file.get({
            storeName: 'notexists',
            name: 'foo.jpg',
          });
          await pipeline(
            stream,
            createWriteStream(filepath),
          );
        },
        (err: any) => {
          assert.equal(err.code, 3);
          assert.equal(err.details, 'not supported store type: notexists');
          return true;
        }
      )
    });
  });

  describe('put()', () => {
    it('should put a file stream to server success', async () => {
      const stream = createReadStream(__filename);
      await client.file.put({
        storeName: 'aliOSS',
        name: 'layotto_js_sdk_unittest.test.ts',
        stream,
      });
    }, 10000);

    it('should put bytes data to server success', async () => {
      await client.file.put({
        storeName: 'aliOSS',
        name: 'layotto_js_sdk_unittest.txt',
        data: Buffer.from('哈哈😄，text still work', 'utf8'),
      });
    }, 10000);

    it.skip('should work with minio', async () => {
      await client.file.put({
        storeName: 'minio',
        name: 'layotto_js_sdk_unittest.txt',
        data: Buffer.from('哈哈😄，text still work', 'utf8'),
      });
    });
  });

  describe('list()', () => {
    it('should list dir success', async () => {
      const { names } = await client.file.list({
        storeName,
        name: 'layotto-js-sdk-local-test',
        metadata: {
          Prefix: 'foo/',
        },
      });
      console.log(names);
      assert(names.length > 0);
    });

    it.skip('should work with minio', async () => {
      const { names } = await client.file.list({
        storeName: 'minio',
        name: 'layotto-js-sdk-local-test',
      });
      console.log(names);
      assert(names.length > 0);
    });
  });

  describe('delete()', () => {
    it('should put bytes data to server and delete it success', async () => {
      const name = `layotto_js_sdk_unittest_need_to_delete_${Date.now()}.txt`;
      await client.file.put({
        storeName,
        name,
        data: Buffer.from('哈哈😄，text still work, delete it now!', 'utf8'),
      });

      await utils.sleep(500);
      const filepath = join(tmpfileDir, name);
      const stream = await client.file.get({
        storeName,
        name,
      });
      assert(stream);
      await pipeline(
        stream,
        createWriteStream(filepath),
      );

      // delete it
      await client.file.delete({
        storeName,
        name,
      });

      // throw error
      await assert.rejects(
        async () => {
          const filepath = join(tmpfileDir, name);
          const stream = await client.file.get({
            storeName,
            name,
          });
          await pipeline(
            stream,
            createWriteStream(filepath),
          );
        },
        (err: any) => {
          // console.error(err);
          assert.equal(err.code, 13);
          assert.match(err.message, /StatusCode=404, ErrorCode=NoSuchKey/);
          return true;
        }
      );
    }, 10000);
  });
});

